src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs

```diff
@@ -1238,16 +1240,44 @@ pub struct QmpSocket {
 }
 
 impl QmpSocket {
-    fn new(_sid: &str) -> Result<Self> {
+    fn new(sid: &str) -> Result<Self> {
+        let sock_path = [KATA_PATH, sid, QMP_SOCKET_FILE].join("/");
+        let listener = UnixListener::bind(sock_path).context("unix listener bind failed.")?;
+
+        let raw_fd = listener.into_raw_fd();
+        clear_cloexec(raw_fd).context("clearing unix listenser O_CLOEXEC failed")?;
+        let sock_file = unsafe { File::from_raw_fd(raw_fd) };
+
         Ok(QmpSocket {
             socket_type: QMP_SOCKET_TYPE.to_owned(),
             protocol: MonitorProtocol::new("qmp"),
-            fd: None,
+            fd: Some(sock_file),
             name: QMP_SOCKET_FILE.to_owned(),
             server: true,
             nowait: true,
         })
     }
+
+    // validate returns true if the QmpSocket structure is valid and complete.
+    #[allow(dead_code)]
+    fn validate(&self) -> bool {
```

Comment:
We don't use `validate()` functions in qemu-rs.  They are used in govmm since its ambition was (initially) much broader so it made sense for its code to be more general.  Where an object set-up depends on outside information like user input it makes sense to validate the result.

In qemu-rs however we're generally in control of our objects' set-up.  (Even where we aren't we prefer detecting problems earlier.)  That's exactly the case for `QmpSocket` - `validate()` is reduced just to restate what the constructor guarantees already and thus seems superfluous.